### PR TITLE
POSIX regular expressions operators

### PR DESCRIPTION
**Pull request by PSeON**
Hello. This pull request adds two POSIX regular expressions operators ("~" and "~*").
https://www.postgresql.org/docs/current/static/functions-matching.html#FUNCTIONS-POSIX-REGEXP
Tests included.
Please let me know what do you think. Thanks.

### CODE REVIEW

**Comment by roji** [comment]
Thanks for submitting this!

The basic work looks good, but there are several important discrepancies between .NET regex and PostgreSQL regex, and it's important for the EF6 provider to provide .NET behavior.

I recommend you take a look at the work I've done on regex in the EFCore provider (see NpgsqlRegexIsMatchTranslator).

File: src/EntityFramework6.Npgsql/SqlGenerators/SqlBaseGenerator.cs [hunk 1]
```diff
@@ -81,7 +81,9 @@ internal Version Version
             {"operator_tsquery_and",Operator.QueryAnd},
             {"operator_tsquery_or",Operator.QueryOr},
             {"operator_tsquery_contains",Operator.QueryContains},
-            {"operator_tsquery_is_contained",Operator.QueryIsContained}
+            {"operator_tsquery_is_contained",Operator.QueryIsContained},
+            {"regex_is_match",Operator.RegexIsMatch},
```
Comment (rwasef1830) [inline src/EntityFramework6.Npgsql/SqlGenerators/SqlBaseGenerator.cs]: I think it would be more consistent to change ```regex_is_match``` to ```operator_regex_is_match```.

File: src/EntityFramework6.Npgsql/SqlGenerators/SqlBaseGenerator.cs [hunk 2]
```diff
@@ -81,7 +81,9 @@ internal Version Version
...
+            {"regex_is_match_case",Operator.RegexIsMatchCase}
```
Comment (rwasef1830) [inline src/EntityFramework6.Npgsql/SqlGenerators/SqlBaseGenerator.cs]: Same thing here

**Comment by PSeON** [comment]
Agree. `operator_regex_is_match` looks more appropriate.

**Comment by PSeON** [comment]
Regular expressions functions now match .NET behaviour. Like in the EFCore provider.

**Comment by roji** [comment]
@PSeON thanks, I promise to review ASAP. In the meantime can you please squash the two commits and push force for a cleaner history?

**Comment by PSeON** [comment]
Done.

**Comment by roji** [comment]
@rwasef1830, if you can help review this and provide more comments that would be greatly appreciated!

File: src/EntityFramework6.Npgsql/NpgsqlTextFunctions.cs [hunk 3]
```diff
@@ -368,5 +369,38 @@ public static string TsRewrite(string query, string target, string substitute)
...
+        /// http://www.postgresql.org/docs/current/static/functions-matching.html#FUNCTIONS-POSIX-REGEXP
+        /// </summary>
+        /// <param name="input">A string to search for a match.</param>
+        /// <param name="pattern">A regular expression pattern to match.</param>
+        /// <returns>
+        /// <see langword="true"/> if <paramref name="input"/> matches <paramref name="pattern"/>,
+        /// otherwise, <see langword="false"/>.
+        /// </returns>
+        [DbFunction("Npgsql", "regex_is_match")]
+        public static bool RegexIsMatch(string input, string pattern)
```
Comment (rwasef1830) [inline src/EntityFramework6.Npgsql/NpgsqlTextFunctions.cs]: I think that if this (and its overload) were named MatchRegex it would be consistent the "Match" method which generates the @@ operator.

File: src/EntityFramework6.Npgsql/NpgsqlTextFunctions.cs [hunk 4]
```diff
@@ -368,5 +369,38 @@ public static string TsRewrite(string query, string target, string substitute)
...
+        /// http://www.postgresql.org/docs/current/static/functions-matching.html#FUNCTIONS-POSIX-REGEXP
+        /// </summary>
+        /// <param name="input">A string to search for a match.</param>
+        /// <param name="pattern">A regular expression pattern to match.</param>
+        /// <param name="options">A bitwise combination of the enumeration values that provide options for matching.</param>
+        /// <returns>
+        /// <see langword="true"/> if <paramref name="input"/> matches <paramref name="pattern"/>,
+        /// otherwise, <see langword="false"/>.
+        /// </returns>
+        [DbFunction("Npgsql", "regex_is_match_options")]
```
Comment (rwasef1830) [inline src/EntityFramework6.Npgsql/NpgsqlTextFunctions.cs]: There is no need to define another name, multiple overloads of the same function can have the same name, and in ```SqlBaseGenerator``` check number of arguments to disambiguate between them. See the functions marked as ```to_tsvector``` in the same file.

File: test/EntityFramework6.Npgsql.Tests/EntityFramework6.Npgsql.Tests.csproj [hunk 5]
```diff
@@ -100,5 +101,8 @@
       <Name>EntityFramework6.Npgsql</Name>
     </ProjectReference>
   </ItemGroup>
+  <ItemGroup>
+    <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
```
Comment (rwasef1830) [inline test/EntityFramework6.Npgsql.Tests/EntityFramework6.Npgsql.Tests.csproj]: This change is completely unrelated to this feature and should be excluded from the pull request (added by visual studio if you have nunit vs integration I think).

File: src/EntityFramework6.Npgsql/NpgsqlTextFunctions.cs [hunk 6]
```diff
@@ -368,5 +369,38 @@ public static string TsRewrite(string query, string target, string substitute)
...
+        /// otherwise, <see langword="false"/>.
+        /// </returns>
+        [DbFunction("Npgsql", "regex_is_match")]
+        public static bool RegexIsMatch(string input, string pattern)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <summary>
+        /// Matches regular expression, case sensitive. Generates the "~" operator.
```
Comment (rwasef1830) [inline src/EntityFramework6.Npgsql/NpgsqlTextFunctions.cs]: This summary is not accurate. It is case sensitive or not based on the passed options, so it shouldn't state a specific sensitivity here. It should state that it follows the .NET Regex semantics and it is translated to the equivalent PostgreSQL expression when executed.

File: src/EntityFramework6.Npgsql/NpgsqlTextFunctions.cs [hunk 7]
```diff
@@ -368,5 +369,38 @@ public static string TsRewrite(string query, string target, string substitute)
...
+        /// http://www.postgresql.org/docs/current/static/functions-matching.html#FUNCTIONS-POSIX-REGEXP
+        /// </summary>
+        /// <param name="input">A string to search for a match.</param>
+        /// <param name="pattern">A regular expression pattern to match.</param>
+        /// <param name="options">A bitwise combination of the enumeration values that provide options for matching.</param>
```
Comment (rwasef1830) [inline src/EntityFramework6.Npgsql/NpgsqlTextFunctions.cs]: This should state that ```RegexOptions.RightToLeft``` and ```RegexOptions.ECMAScript``` are unsupported to prevent consumers of this code from getting surprised. (As-is, people would find out at runtime or by reading the source).

**Comment by rwasef1830** [comment]
@PSeON in addition to my notes on the commit, I find the behavior difference between the overload without RegexOptions and the one with to be slightly misleading.

There are several sets of semantics exposed in a confusing way here: Operator "~" vs "~*" and behavior between both overloads (with and without .NET RegexOptions) and lining up with PostgreSQL semantics.

According to [postgresql docs](https://www.postgresql.org/docs/9.5/static/functions-matching.html#FUNCTIONS-POSIX-REGEXP):
>An ARE can begin with embedded options: a sequence (?xyz) (where xyz is one or more alphabetic characters) specifies options affecting the rest of the RE. These options override any previously determined options — in particular, they can override the case-sensitivity behavior implied by a regex operator, or the flags parameter to a regex function. 

You should not have Npgsql use both operators and in the case of one of them override the meaning of the operator with embedded options. To match the .NET semantics (which is case-sensitive by default unless overridden by RegexOptions), both functions should emit the case-sensitive version of the PostgreSQL operator. 

In the case of the second one, the meaning of the operator will be overridden by the RegexOptions, and in case of ```RegexOptions.None``` (or calling the overload that doesn't have a RegexOptions parameter), it should behave as .NET would behave (and both methods should behave the same).

There should be also testcases that confirm this behavior.

@roji What do you think about this and my notes ?

File: test/EntityFramework6.Npgsql.Tests/PatternMatchingTests.cs [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+                {
+                    var results = (from b in context.Blogs
+                                   where NpgsqlTextFunctions.RegexIsMatch(b.Name, "Some pattern", RegexOptions.ECMAScript)
+                                   select b.Name).ToList();
+                });
+            }
+        }
+
+        [Test]
+        public void RegexIsMatchOptions()
```
Comment (rwasef1830) [inline test/EntityFramework6.Npgsql.Tests/PatternMatchingTests.cs]: I think this could be clearer / more readable using [TestCase].

**Comment by PSeON** [comment]
@rwasef1830 thanks for the comments. I am very grateful to you for your help. I will fix these problems ASAP.

**Comment by PSeON** [comment]
Finished implementation of recommendations. Please review this version. Thanks.

File: src/EntityFramework6.Npgsql/SqlGenerators/SqlBaseGenerator.cs [hunk 9]
```diff
@@ -1160,6 +1167,77 @@ VisitedExpression VisitFunction(EdmFunction function, IList<DbExpression> args,
...
+            if (options == RegexOptions.Singleline)
+            {
+                return OperatorExpression.Build(
+                    Operator.RegexMatchCase,
+                    _useNewPrecedences,
+                    args[0].Accept(this),
+                    args[1].Accept(this));
+            }
+
+            string flags = "(?";
```
Comment (roji) [inline src/EntityFramework6.Npgsql/SqlGenerators/SqlBaseGenerator.cs]: Please use a StringBuilder rather than string concatenation

**Comment by roji** [comment]
I've taken a quick look (sorry, don't have much spare time at the moment), it's after @PSeON's latest modifications so @rwasef1830, I'm not sure exactly what you were referring to before - although what you say sounds right. The code as it is now seems to correspond pretty well to what I wrote for EF Core. I remember I spent some time understanding exactly how PostgreSQL does regexp and how .NET does them, and I think everything should be OK.

@rwasef1830, if you thing any discrepancy still remains please let us know (I'll fix the EF Core implementation too if necessary), otherwise I'm OK for merging (modulu the nit-picking StringBuilder comment). @rwasef1830 I'll wait for your confirmation before doing so.

File: src/EntityFramework6.Npgsql/SqlGenerators/SqlBaseGenerator.cs [hunk 10]
```diff
@@ -81,7 +82,9 @@ internal Version Version
             {"operator_tsquery_and",Operator.QueryAnd},
             {"operator_tsquery_or",Operator.QueryOr},
             {"operator_tsquery_contains",Operator.QueryContains},
-            {"operator_tsquery_is_contained",Operator.QueryIsContained}
+            {"operator_tsquery_is_contained",Operator.QueryIsContained},
+            {"operator_regex_match",Operator.RegexMatch},
+            {"operator_regex_match_case",Operator.RegexMatchCase}
```
Comment (rwasef1830) [inline src/EntityFramework6.Npgsql/SqlGenerators/SqlBaseGenerator.cs]: ```operator_regex_match_case``` should be removed (it contradicts the method documentation).

File: src/EntityFramework6.Npgsql/SqlGenerators/SqlBaseGenerator.cs [hunk 11]
```diff
@@ -81,7 +82,9 @@ internal Version Version
             {"operator_tsquery_and",Operator.QueryAnd},
             {"operator_tsquery_or",Operator.QueryOr},
             {"operator_tsquery_contains",Operator.QueryContains},
-            {"operator_tsquery_is_contained",Operator.QueryIsContained}
+            {"operator_tsquery_is_contained",Operator.QueryIsContained},
+            {"operator_regex_match",Operator.RegexMatch},
```
Comment (rwasef1830) [inline src/EntityFramework6.Npgsql/SqlGenerators/SqlBaseGenerator.cs]: Please rename ```operator_regex_match_case``` to ```operator_regex_match``` (and remove the other one). This implementation should be only using 1 operator and that's the case sensitive one (and the behavior is then overriden with modifiers according to RegexOptions).

File: src/EntityFramework6.Npgsql/SqlGenerators/SqlBaseGenerator.cs [hunk 12]
```diff
@@ -1160,6 +1167,77 @@ VisitedExpression VisitFunction(EdmFunction function, IList<DbExpression> args,
...
+            if (args.Count == 3)
+            {
+                var optionsExpression = args[2] as DbConstantExpression;
+                if (optionsExpression == null)
+                    throw new NotSupportedException("Options must be constant expression.");
+
+                options = (RegexOptions)optionsExpression.Value;
+            }
+
+            const RegexOptions UnsupportedRegexOptions = RegexOptions.RightToLeft | RegexOptions.ECMAScript;
```
Comment (rwasef1830) [inline src/EntityFramework6.Npgsql/SqlGenerators/SqlBaseGenerator.cs]: Nitpick: I find this const a bit out of place, it is not referred to anywhere else and the exception thrown directly below it clarifies it, just inline the variable.

File: src/EntityFramework6.Npgsql/SqlGenerators/VisitedExpression.cs [hunk 13]
```diff
@@ -844,6 +844,8 @@ internal enum UnaryTypes {
         public static readonly Operator QueryNegate = new Operator("!!", 10, 8, UnaryTypes.Prefix, true);
         public static readonly Operator QueryContains = new Operator("@>", 10, 8);
         public static readonly Operator QueryIsContained = new Operator("<@", 10, 8);
+        public static readonly Operator RegexMatch = new Operator("~*", 10, 8);
+        public static readonly Operator RegexMatchCase = new Operator("~", 10, 8);
```
Comment (rwasef1830) [inline src/EntityFramework6.Npgsql/SqlGenerators/VisitedExpression.cs]: Please rename RegexMatchCase to just RegexMatch and remove the insensitive one (it is unused).

File: test/EntityFramework6.Npgsql.Tests/PatternMatchingTests.cs [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        [TestCase("blog", "blog", "BLOG")]
+
+        // The "^" and "$" language elements match the beginning and end of the input string
+        [TestCase("^blog$", "blog", "some \nblog\n name")]
+
+        // The "." language element matches every character except "\n"
+        [TestCase("some .* name", "some blog name", "some \n name")]
+
+        // Any white space in a regular expression pattern is interpreted as a literal space character
+        [TestCase("some blog name", "some blog name", "someblogname")]
```
Comment (rwasef1830) [inline test/EntityFramework6.Npgsql.Tests/PatternMatchingTests.cs]: The empty spaces between the attributes are not needed, also the comments don't add value. It is better to use the ```TestName``` property of the TestCase attribute and set a really short descriptive name (eg: ```Case-sensitive``` for the first, ```^ and $ match beginning and end```, ```. matches all except \\n```, ```Whitespace not ignored in pattern```) so that it shows up in the test runner instead of the test parameters.

I also think the [Test] attribute should be added on the top of all the [TestCase] attributes (some test runners might need it).

File: test/EntityFramework6.Npgsql.Tests/PatternMatchingTests.cs [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+            Assert.That(pgResults[0], Is.EqualTo(matchingInput));
+        }
+
+        [TestCase(RegexOptions.RightToLeft)]
+        [TestCase(RegexOptions.ECMAScript)]
+        public void MatchRegex_NotSupportedOption(RegexOptions options)
+        {
+            using (var context = new BloggingContext(ConnectionString))
+            {
+                Assert.Catch(() =>
```
Comment (rwasef1830) [inline test/EntityFramework6.Npgsql.Tests/PatternMatchingTests.cs]: Please assert the exception type (```NotSupportedException```)

**Comment by rwasef1830** [comment]
@PSeON Thanks for your work! In addition to my latest comments, I think you should add a test for negative matching (maybe a second query in the same tests that selects the mismatching input).

@roji Will negating this query execute and be treated the same way as using the "not regex match" operator in PostgreSQL or does this need to be separately implemented as well ?

**Comment by roji** [comment]
@rwasef1830, good question - it's worth testing. I'm guessing that without any extra handling, a negative regex match should render something like `NOT (x ~ y)`, which is correct but maybe a tiny bit ugly. I'm not sure it's worth the extra effort to render `x !~ y`...

**Comment by roji** [comment]
@rwasef1830 and thanks for the very valuable and through reviewing!

**Comment by rwasef1830** [comment]
@PSeON I forgot to mention, please add a unit test that very explicitly makes sure that using ```MatchRegex(string, string)``` and ```MatchRegex(string, string, RegexOptions.None)``` produce exactly the same results.

**Comment by PSeON** [comment]
@rwasef1830 This [code](https://github.com/npgsql/EntityFramework6.Npgsql/pull/41/commits/7c51f05aedefbfab4c278859cad92e5f4ae5aa39#diff-af93318afb1f0b9301e0f3ce512cce61R75) (line 75) checks that both methods produce same results. Am I wrong?

**Comment by PSeON** [comment]
@roji Checked query rendering.
`NpgsqlTextFunctions.MatchRegex(b.Name, pattern)` renders `"Extent1"."Name" ~ (E'(?p)' || $1) = TRUE`
and
`!NpgsqlTextFunctions.MatchRegex(b.Name, pattern)` renders `"Extent1"."Name" ~ (E'(?p)' || $1) != TRUE`
Of course it is possible to add proposed optimization, but I think it is out of scope of this PR.

**Comment by rwasef1830** [comment]
@PSeON You're right I missed that. Maybe the variable could be clarified, perhaps you could rename the ```pgResult``` and ```pgResultOpt``` to ```pgMatchResult``` and ```pgMatchWithOptionsResult``` so that they line up with your already existing ```netMatchResult``` and the intention is clearer.

(it is better to have the variables clearly named than to write clarifying comments).

**Comment by roji** [comment]
Agree on the negation, the only impact is slightly better SQL readability which really isn't that important here.

**Comment by roji** [comment]
@rwasef1830, whenever you're OK let me know and I'll merge.

**Comment by PSeON** [comment]
Added proposed improvements.

File: test/EntityFramework6.Npgsql.Tests/PatternMatchingTests.cs [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+namespace EntityFramework6.Npgsql.Tests
+{
+    class PatternMatchingTests : EntityFrameworkTestBase
+    {
+        [Test]
+        [TestCase("blog", "blog", "BLOG", TestName = "MatchRegex case-sensitive")]
+        [TestCase("^blog$", "blog", "some \nblog\n name", TestName = "MatchRegex ^ and $ match beginning and end")]
+        [TestCase("some .* name", "some blog name", "some \n name", TestName = "MatchRegex . matches all except \\n")]
+        [TestCase("some blog name", "some blog name", "someblogname", TestName = "MatchRegex whitespace not ignored in pattern")]
+        public void MatchRegex(string pattern, string matchingInput, string mismatchingInput)
```
Comment (rwasef1830) [inline test/EntityFramework6.Npgsql.Tests/PatternMatchingTests.cs]: You can remove MatchRegex from test TestName, in the unit test runner the test names will appear under the method name so no need to mention it again in the test name.

File: test/EntityFramework6.Npgsql.Tests/PatternMatchingTests.cs [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+            Assert.That(pgMismatchWithOptionsResults.Count, Is.EqualTo(1));
+            Assert.That(pgMismatchWithOptionsResults[0], Is.EqualTo(mismatchingInput));
+        }
+
+        [Test]
+        [TestCase(RegexOptions.IgnoreCase, "some", "SOME", "placeholder", TestName = "MatchRegexOptions IgnoreCase")]
+        [TestCase(RegexOptions.IgnorePatternWhitespace, "s o m e", "some", "s o m e", TestName = "MatchRegexOptions IgnorePatternWhitespace")]
+        [TestCase(RegexOptions.Multiline, "^blog$", "some \nblog\n name", "placeholder", TestName = "MatchRegexOptions Multiline")]
+        [TestCase(RegexOptions.Singleline, "some .* name", "some \n name", "placeholder", TestName = "MatchRegexOptions Singleline")]
+        public void MatchRegexOptions(RegexOptions options, string pattern, string matchingInput, string mismatchingInput)
```
Comment (rwasef1830) [inline test/EntityFramework6.Npgsql.Tests/PatternMatchingTests.cs]: Same here, remove "MatchRegexOptions" from the ```TestName``` property value.

**Comment by rwasef1830** [comment]
@PSeON With the minor final changes I proposed to improve the test-case readability, if @roji has no further comments, I think this is OK for merge.

**Comment by PSeON** [comment]
Done.

**Comment by roji** [comment]
Thanks to both of you for this!

**Comment by PSeON** [comment]
@roji, @rwasef1830 Thanks!

**Comment by danbopes** [comment]
Any chance we can get a new version pushed with these changes?

**Comment by roji** [comment]
@danbopes sorry a new version hasn't been released in so long. @rwasef1830, what do you think? As I'm really unlikely to do any work on the EF6 provider anytime soon (have little time for Npgsql in general at the moment), can you take a look at other issues/PRs and tell me if you intend to do any work? If not I can do a release soon.

**Comment by rwasef1830** [comment]
I've been waiting to get a chance to work on them, but didn't get any time I'm afraid :-(

**Comment by roji** [comment]
@rwasef1830 so what do you think, should I release or do you want a bit more time?

**Comment by rwasef1830** [comment]
@roji these 2 pull requests seem almost ready, I'll do a final pass on them today: https://github.com/npgsql/EntityFramework6.Npgsql/pull/75 and https://github.com/npgsql/EntityFramework6.Npgsql/pull/74.

I think that a few days to merge both these and then confirm / investigate https://github.com/npgsql/EntityFramework6.Npgsql/pull/46

**Comment by roji** [comment]
Great, I'll hold off then. Will probably have a bit more time for a release next Sunday, hopefully you can get everything done by then.
